core/chain_pow.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2015 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"runtime"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/pow"
+)
+
+// nonceCheckResult contains the result of a nonce verification.
+type nonceCheckResult struct {
+	index int  // Index of the item verified from an input array
+	valid bool // Result of the nonce verification
+}
+
+// verifyHeaderNonces starts a concurrent header nonce verification, returning a
+// quit channel to abort the operations and a results channel for retrieving the
+// async verifications.
+func verifyHeaderNonces(checker pow.PoW, headers []*types.Header) (chan<- struct{}, <-chan nonceCheckResult) {
+	items := make([]pow.Block, len(headers))
+	for i, header := range headers {
+		items[i] = types.NewBlockWithHeader(header)
+	}
+	return verifyNonces(checker, items)
+}
+
+// verifyBlockNonces starts a concurrent block nonce verification, returning a
+// quit channel to abort the operations and a results channel for retrieving the
+// async verifications.
+func verifyBlockNonces(checker pow.PoW, blocks []*types.Block) (chan<- struct{}, <-chan nonceCheckResult) {
```

Comment:
hmm, if ```verifyBlockNonces``` is removed and the caller just calls ```verifyNonces```, will it automatically make interface conversion from []*types.Block to []pow.Block? Otherwise maybe just call them ```verifyBlocks``` and ```verifyHeaders``` - the actual PoW uses several block header values in verification: mixhash, nonce, difficulty, block time, block hash. Which is why the PoW interface takes a full block.